src/wasm/wasm-type.cpp

```diff
@@ -1260,6 +1299,15 @@ Canonicalizer::Canonicalizer(TypeBuilder& builder) : builder(builder) {
         break;
     }
   }
+
+  // Now that all uses of Types and non-recursive HeapTypes have been
+  // canonicalized, move recursive HeapTypes to the global store so that they
+  // will outlive the TypeBuilder without their IDs changing.
+  for (auto& entry : builder.impl->entries) {
+    if (entry.info->recursive) {
+      globalHeapTypeStore.insertWithoutCanonicalizing(std::move(entry.info));
```

Comment:
Can we do the same for all types? I understand this is needed for recursive types, but if it's possible for non-recursive ones as well that could be simpler.

Reply:
If we end up partially or wholly switching to nominal types, it would make sense to move ownership of the nominal types from the canonicalizing global stores to the individual `Module`s. But for structural types, deduplication is a matter of correctness. If the proposal keeps fully structural typing, we will have to implement canonicalization of recursive types at some point as well.

Reply:
Oh, I see, I didn't get that before. So this does not canonicalize recursive types despite it being technically incorrect, but it may be good enough for now? (Why is it good enough for now?)

Reply:
This should be good enough for now because producers will generally emit one type declaration per source type and not deduplicate or mix the two. Because producers don’t treat the types as interchangeable, it’s ok that Binaryen doesn’t either. Another piece is that for now neither Binaryen nor the producers are dealing with systems of multiple modules. If they were, it would be important for Binaryen to recognize when the same type is declared in multiple modules, which would require properly canonicalizing recursive types.